### PR TITLE
Sniffer

### PR DESCRIPTION
**Pull request by patricklee2**
update sniffers to use linuxfxversion

https://msazure.visualstudio.com/One/_git/AAPT-Antares-Websites/pullrequest/2421919?_a=overview

### CODE REVIEW

File: Kudu.Core/Deployment/Oryx/AppServiceOryxArguments.cs [hunk 1]
```diff
@@ -180,13 +168,23 @@ public string GenerateOryxBuildCommand(DeploymentContext context)
...
+                    {
+                        OryxArgumentsHelper.AddOryxBuildCommand(args, source: context.RepositoryPath, destination: context.OutputPath);
+                    }
+                    break;
             }
 
             // Version
             switch (Language)
             {
-                case Framework.None:
```
Comment (patricklee2) [inline Kudu.Core/Deployment/Oryx/AppServiceOryxArguments.cs]: already has a default

File: Kudu.Core/Deployment/Generator/SiteBuilderFactory.cs [hunk 2]
```diff
@@ -178,33 +179,37 @@ public ISiteBuilder CreateBuilder(ITracer tracer, ILogger logger, IDeploymentSet
 
         private ISiteBuilder ResolveNonAspProject(string repositoryRoot, string projectPath, IDeploymentSettingsManager perDeploymentSettings)
         {
+            string strFramework = System.Environment.GetEnvironmentVariable(OryxBuildConstants.OryxEnvVars.FrameworkSetting);
+            Framework framework = SupportedFrameworks.ParseLanguage(strFramework);
```
Comment (patricklee2) [inline Kudu.Core/Deployment/Generator/SiteBuilderFactory.cs]: use preexisting parser to parse framework env var

File: Kudu.Core/Deployment/Oryx/SupportedLanguages.cs [hunk 3]
```diff
@@ -8,7 +8,9 @@ public enum Framework
         NodeJs,
         Python,
         DotNETCore,
-        PHP
+        PHP,
```
Comment (patricklee2) [inline Kudu.Core/Deployment/Oryx/SupportedLanguages.cs]: add ruby, go

File: Kudu.Core/Deployment/Oryx/AppServiceOryxArguments.cs [hunk 4]
```diff
@@ -57,9 +57,6 @@ private void SetLanguageOptions()
         {
             switch(Language)
             {
-                case Framework.None:
```
Comment (patricklee2) [inline Kudu.Core/Deployment/Oryx/AppServiceOryxArguments.cs]: replace with default

File: Kudu.Core/Deployment/Oryx/AppServiceOryxArguments.cs [hunk 5]
```diff
@@ -116,18 +116,6 @@ public string GenerateOryxBuildCommand(DeploymentContext context)
             // Language
             switch (Language)
             {
-                case Framework.None:
```
Comment (patricklee2) [inline Kudu.Core/Deployment/Oryx/AppServiceOryxArguments.cs]: replace with default

File: Kudu.Core/Deployment/Oryx/OryxArguments.cs [hunk 6]
```diff
@@ -53,9 +53,6 @@ private void SetLanguageOptions()
         {
             switch(Language)
             {
-                case Framework.None:
```
Comment (patricklee2) [inline Kudu.Core/Deployment/Oryx/OryxArguments.cs]: replace with default

File: Kudu.Core/Deployment/Oryx/OryxArguments.cs [hunk 7]
```diff
@@ -102,9 +102,6 @@ public string GenerateOryxBuildCommand(DeploymentContext context)
             // Language
             switch (Language)
             {
-                case Framework.None:
```
Comment (patricklee2) [inline Kudu.Core/Deployment/Oryx/OryxArguments.cs]: replace with default

File: Kudu.Core/Deployment/Generator/SiteBuilderFactory.cs [hunk 8]
```diff
@@ -178,33 +179,57 @@ public ISiteBuilder CreateBuilder(ITracer tracer, ILogger logger, IDeploymentSet
...
+                case Framework.PHP:
+                    return new PHPSiteBuilder(_environment, perDeploymentSettings, _propertyProvider, repositoryRoot, projectPath);
 
-            return new BasicBuilder(_environment, perDeploymentSettings, _propertyProvider, repositoryRoot, projectPath);
+                default:
+                    if (FunctionAppHelper.LooksLikeFunctionApp())
+                    {
+                        return new FunctionBasicBuilder(_environment, perDeploymentSettings, _propertyProvider, repositoryRoot, projectPath);
+                    }
+                    else if (IsRubySite(sourceProjectPath))
```
Comment (patricklee2) [inline Kudu.Core/Deployment/Generator/SiteBuilderFactory.cs]: - default to sniffers if linuxfxversion is not set, customer may be using a custom image, and relying on sniffers
- place ruby before node

File: Kudu.Core/Deployment/Generator/SiteBuilderFactory.cs [hunk 9]
```diff
@@ -178,33 +179,57 @@ public ISiteBuilder CreateBuilder(ITracer tracer, ILogger logger, IDeploymentSet
...
-            else if (IsNodeSite(sourceProjectPath))
-            {
-                return new NodeSiteBuilder(_environment, perDeploymentSettings, _propertyProvider, repositoryRoot, projectPath);
-            }
-            else if (IsPythonSite(sourceProjectPath))
-            {
-                return new PythonSiteBuilder(_environment, perDeploymentSettings, _propertyProvider, repositoryRoot, projectPath);
-            }
-            else if (IsGoSite(sourceProjectPath))
+            switch (framework)
```
Comment (patricklee2) [inline Kudu.Core/Deployment/Generator/SiteBuilderFactory.cs]: @sanchitmehta It shouldnt be possible to the a blessed image for the linuxfxversion, and the customer using the sniffer to detect another framework,  The sniffed runtime would not be present.

File: Kudu.Core/Deployment/Generator/SiteBuilderFactory.cs (hunk 8 above)
Comment (sanchitmehta) [reply Kudu.Core/Deployment/Generator/SiteBuilderFactory.cs]: Before switching to other builders FunctionApp should still get the first precedence.

Comment (sanchitmehta) [reply Kudu.Core/Deployment/Generator/SiteBuilderFactory.cs]: Why do we have all the switch cases in default part again

Comment (patricklee2) [reply Kudu.Core/Deployment/Generator/SiteBuilderFactory.cs]: If we cant choose based on the linuxfxversion, we sniff the files

**Pull request closed without merging**